### DISCUSSION TITLE
Using the calculator tool and the json format ticks (```json)

### DISCUSSION THREAD

**Discussion by aleixdorca** [open-webui/open-webui#3522] [Q&A]
Hi all,

I've been trying Tools in version 0.3.6. I've created the Calculator tool copying it from the openwebui repository and I have enabled it in the model and in the current Chat. So far so good.

The thing is that the tool is seldomly executed. It looks like openwebui generates a json to pass to the python code. It uses the current model to do so. Most of the time the response includes ```json at the beginning and ``` at the end.

If this is the case the Tool is not executed.

Is there anyway to force the LLM not to return these code markup? I'm pretty sure it may have something to do with the loaded model, maybe each is different in generating the context. In this case I'm using Gemma2-9b-q8_0.

Cheers.

**Comment by aleixdorca** [comment]
Ok, I'll follow my own discussion. This was driving me crazy.

I've traced the code to the place where the response from Gemma2, after using the tool, was processed.

First I tried rewriting the prompt that makes the LLM choose a tool, adding the typical: "don't add ```json at the beginning". It did not work as expected. Gemma2 continues to add the preamble whenever it feels.

Then I've added the following code (don't want to create a PR for something so specific):

/app/backend/main.py:

```
import re
```

And then, at line 352 (from version 0.3.8):

```
content = re.sub(r"```json", "", content)
content = re.sub(r"```", "", content)
```

I am sure there's a better way to do so, maybe by trying to load the json content and try/catch the error...

This improves the tools usage from 30%-ish to more than 90% of the times doing it correctly.

Just my 2 cents.

Cheers.

**Comment by aleixdorca** [comment]
Here is a different solution:

In this case I use a recursive regex definition of a valid JSON (from: [https://stackoverflow.com/questions/2583472/regex-to-validate-json](https://stackoverflow.com/questions/2583472/regex-to-validate-json)), and extract the first valid string that contains the structure needed to use the tools.

```
import regex

content = '```json\n{ "name": "functionName", "parameters": { "key": "value" } }\n```'
pattern = r'(?(DEFINE)(?<json>(?>\s*(?&object)\s*|\s*(?&array)\s*))(?<object>(?>\{\s*(?>(?&pair)(?>\s*,\s*(?&pair))*)?\s*\}))(?<pair>(?>(?&STRING)\s*:\s*(?&value)))(?<array>(?>\[\s*(?>(?&value)(?>\s*,\s*(?&value))*)?\s*\]))(?<value>(?>true|false|null|(?&STRING)|(?&NUMBER)|(?&object)|(?&array)))(?<STRING>(?>"(?>\\(?>["\\\/bfnrt]|u[a-fA-F0-9]{4})|[^"\\\0-\x1F\x7F]+)*"))(?<NUMBER>(?>-?(?>0|[1-9][0-9]*)(?>\.[0-9]+)?(?>[eE][+-]?[0-9]+)?))) (?&json)'

content = regex.sub("\n", " ", content) # replace newlines with spaces
match = regex.search(pattern, content) # search for occurences of json strings
print(match.group(0).strip()) # prints { "name": "functionName", "parameters": { "key": "value" } }
```

Could it work? Is it feasible?

Thanks.

**Reply by flyfox666** [reply]
hi，also face the probelm of jason, just want to know how to use follow as you said in openwebui ,shall I need to change de main.py?

import regex

content = '```json\n{ "name": "functionName", "parameters": { "key": "value" } }\n```'
pattern = r'(?(DEFINE)(?<json>(?>\s*(?&object)\s*|\s*(?&array)\s*))(?<object>(?>\{\s*(?>(?&pair)(?>\s*,\s*(?&pair))*)?\s*\}))(?<pair>(?>(?&STRING)\s*:\s*(?&value)))(?<array>(?>\[\s*(?>(?&value)(?>\s*,\s*(?&value))*)?\s*\]))(?<value>(?>true|false|null|(?&STRING)|(?&NUMBER)|(?&object)|(?&array)))(?<STRING>(?>"(?>\\(?>["\\\/bfnrt]|u[a-fA-F0-9]{4})|[^"\\\0-\x1F\x7F]+)*"))(?<NUMBER>(?>-?(?>0|[1-9][0-9]*)(?>\.[0-9]+)?(?>[eE][+-]?[0-9]+)?))) (?&json)'

content = regex.sub("\n", " ", content) # replace newlines with spaces
match = regex.search(pattern, content) # search for occurences of json strings
print(match.group(0).strip()) # prints { "name": "functionName", "parameters": { "key": "value" } }

**Reply by aleixdorca** [reply]
That is correct. I edited the main.py file and added the import at the beginning.

The rest of the lines go in this block of code (at lines 349-352 of version 0.3.10):

```
        # Parse the function response
        print(f"content: {content}")
        result = json.loads(content)
        print(result)
```

becoming:

```
        # Parse the function response
        print(f"content: {content}")
        pattern = r'(?(DEFINE)(?(?>\s*(?&object)\s*|\s*(?&array)\s*))(?(?>{\s*(?>(?&pair)(?>\s*,\s*(?&pair)))?\s}))(?(?>(?&STRING)\s*:\s*(?&value)))(?(?>[\s*(?>(?&value)(?>\s*,\s*(?&value)))?\s]))(?(?>true|false|null|(?&STRING)|(?&NUMBER)|(?&object)|(?&array)))(?(?>"(?>\(?>["\/bfnrt]|u[a-fA-F0-9]{4})|[^"\\\0-\x1F\x7F]+)"))(?(?>-?(?>0|[1-9][0-9])(?>.[0-9]+)?(?>[eE][+-]?[0-9]+)?))) (?&json)'

        content = regex.sub("\n", " ", content) # replace newlines with spaces
        content = regex.search(pattern, content) 
        result = json.loads(content)
        print(result)
```

I understand this is far from perfect. I've tried with other models like mistral v3 and function calling woks like a charm right out of the box... oh, well.

Let me know if you have any further questions.

**Reply by flyfox666** [reply]
Really Thanks. 
Actually, my main issue is with the stock report tool. Whether can I also apply your code and approach to solve it, as follows:

"""
title: Stock Market Helper
description: A comprehensive stock analysis tool that gathers data from Finnhub API and compiles a detailed report.
author: Pyotr Growpotkin
author_url: https://github.com/christ-offer/
github: https://github.com/christ-offer/open-webui-tools
funding_url: https://github.com/open-webui
version: 0.0.9
license: MIT
requirements: finnhub-python
"""

import finnhub
import requests
import aiohttp
import asyncio
from transformers import pipeline, AutoTokenizer, AutoModelForSequenceClassification
import torch
from bs4 import BeautifulSoup
from pydantic import BaseModel
from datetime import datetime, timedelta
from typing import (
    Dict,
    Any,
    List,
    Union,
    Generator,
    Iterator,
    Tuple,
    Optional,
    Callable,
    Awaitable,
)
from functools import lru_cache


def _format_date(date: datetime) -> str:
    """Helper function to format date for Finnhub API"""
    return date.strftime("%Y-%m-%d")


# Caching for expensive operations
@lru_cache(maxsize=128)
def _get_sentiment_model():
    model_name = "mrm8488/distilroberta-finetuned-financial-news-sentiment-analysis"
    tokenizer = AutoTokenizer.from_pretrained(model_name)
    model = AutoModelForSequenceClassification.from_pretrained(model_name)
    return tokenizer, model


def _get_basic_info(client: finnhub.Client, ticker: str) -> Dict[str, Any]:
    """
    Fetch comprehensive company information from Finnhub API.
    """
    profile = client.company_profile2(symbol=ticker)
    basic_financials = client.company_basic_financials(ticker, "all")
    peers = client.company_peers(ticker)

    return {"profile": profile, "basic_financials": basic_financials, "peers": peers}


def _get_current_price(client: finnhub.Client, ticker: str) -> Dict[str, float]:
    """
    Fetch current price and daily change from Finnhub API.
    """
    quote = client.quote(ticker)
    return {
        "current_price": quote["c"],
        "change": quote["dp"],
        "change_amount": quote["d"],
        "high": quote["h"],
        "low": quote["l"],
        "open": quote["o"],
        "previous_close": quote["pc"],
    }


def _get_company_news(client: finnhub.Client, ticker: str) -> List[Dict[str, str]]:
    """
    Fetch recent news articles about the company from Finnhub API.
    Returns a list of dictionaries containing news item details.
    """
    end_date = datetime.now()
    start_date = end_date - timedelta(days=7)
    news = client.company_news(ticker, _format_date(start_date), _format_date(end_date))

    news_items = news[:10]  # Get the first 10 news items

    return [{"url": item["url"], "title": item["headline"]} for item in news_items]


async def _async_web_scrape(session: aiohttp.ClientSession, url: str) -> str:
    """
    Scrape and process a web page using r.jina.ai

    :param session: The aiohttp ClientSession to use for the request.
    :param url: The URL of the web page to scrape.
    :return: The scraped and processed content without the Links/Buttons section, or an error message.
    """
    jina_url = f"https://r.jina.ai/{url}"

    headers = {
        "X-No-Cache": "true",
        "X-With-Images-Summary": "true",
        "X-With-Links-Summary": "true",
    }

    try:
        async with session.get(jina_url, headers=headers) as response:
            response.raise_for_status()
            content = await response.text()

        # Extract content and remove Links/Buttons section as its too many tokens
        links_section_start = content.rfind("Images:")
        if links_section_start != -1:
            content = content[:links_section_start].strip()

        return content

    except aiohttp.ClientError as e:
        return f"Error scraping web page: {str(e)}"


# Asynchronous sentiment analysis
async def _async_sentiment_analysis(content: str) -> Dict[str, Union[str, float]]:
    tokenizer, model = _get_sentiment_model()

    inputs = tokenizer(content, return_tensors="pt", truncation=True, max_length=512)

    with torch.no_grad():
        outputs = model(**inputs)

    probabilities = torch.nn.functional.softmax(outputs.logits, dim=-1)
    sentiment_scores = probabilities.tolist()[0]

    # Update sentiment labels to match the new model's output
    sentiments = ["Neutral", "Positive", "Negative"]
    sentiment = sentiments[sentiment_scores.index(max(sentiment_scores))]

    confidence = max(sentiment_scores)

    return {"sentiment": sentiment, "confidence": confidence}


# Asynchronous data gathering
async def _async_gather_stock_data(
    client: finnhub.Client, ticker: str
) -> Dict[str, Any]:
    basic_info = _get_basic_info(client, ticker)
    current_price = _get_current_price(client, ticker)
    news_items = _get_company_news(client, ticker)

    async with aiohttp.ClientSession() as session:
        scrape_tasks = [_async_web_scrape(session, item["url"]) for item in news_items]
        contents = await asyncio.gather(*scrape_tasks)

    sentiment_tasks = [
        _async_sentiment_analysis(content) for content in contents if content
    ]
    sentiments = await asyncio.gather(*sentiment_tasks)

    sentiment_results = [
        {
            "url": news_items[i]["url"],
            "title": news_items[i]["title"],
            # "content": contents[i][:500] + "..." if contents[i] and len(contents[i]) > 500 else contents[i],
            "sentiment": sentiment["sentiment"],
            "confidence": sentiment["confidence"],
        }
        for i, sentiment in enumerate(sentiments)
        if contents[i]
    ]

    return {
        "basic_info": basic_info,
        "current_price": current_price,
        "sentiments": sentiment_results,
    }


def _compile_report(data: Dict[str, Any]) -> str:
    """
    Compile gathered data into a comprehensive structured report.
    """
    profile = data["basic_info"]["profile"]
    financials = data["basic_info"]["basic_financials"]
    metrics = financials["metric"]
    peers = data["basic_info"]["peers"]
    price_data = data["current_price"]

    report = f"""
    Comprehensive Stock Analysis Report for {profile['name']} ({profile['ticker']})

    Basic Information:
    Industry: {profile.get('finnhubIndustry', 'N/A')}
    Market Cap: {profile.get('marketCapitalization', 'N/A'):,.0f} M USD
    Share Outstanding: {profile.get('shareOutstanding', 'N/A'):,.0f} M
    Country: {profile.get('country', 'N/A')}
    Exchange: {profile.get('exchange', 'N/A')}
    IPO Date: {profile.get('ipo', 'N/A')}

    Current Trading Information:
    Current Price: ${price_data['current_price']:.2f}
    Daily Change: {price_data['change']:.2f}% (${price_data['change_amount']:.2f})
    Day's Range: ${price_data['low']:.2f} - ${price_data['high']:.2f}
    Open: ${price_data['open']:.2f}
    Previous Close: ${price_data['previous_close']:.2f}

    Key Financial Metrics:
    52 Week High: ${financials['metric'].get('52WeekHigh', 'N/A')}
    52 Week Low: ${financials['metric'].get('52WeekLow', 'N/A')}
    P/E Ratio: {financials['metric'].get('peBasicExclExtraTTM', 'N/A')}
    EPS (TTM): ${financials['metric'].get('epsBasicExclExtraItemsTTM', 'N/A')}
    Return on Equity: {financials['metric'].get('roeRfy', 'N/A')}%
    Debt to Equity: {financials['metric'].get('totalDebtToEquityQuarterly', 'N/A')}
    Current Ratio: {financials['metric'].get('currentRatioQuarterly', 'N/A')}
    Dividend Yield: {financials['metric'].get('dividendYieldIndicatedAnnual', 'N/A')}%

    Peer Companies: {', '.join(peers[:5])}

    Detailed Financial Analysis:

    1. Valuation Metrics:
    P/E Ratio: {metrics.get('peBasicExclExtraTTM', 'N/A')}
    - Interpretation: {'High (may be overvalued)' if metrics.get('peBasicExclExtraTTM', 0) > 25 else 'Moderate' if 15 <= metrics.get('peBasicExclExtraTTM', 0) <= 25 else 'Low (may be undervalued)'}

    P/B Ratio: {metrics.get('pbQuarterly', 'N/A')}
    - Interpretation: {'High' if metrics.get('pbQuarterly', 0) > 3 else 'Moderate' if 1 <= metrics.get('pbQuarterly', 0) <= 3 else 'Low'}

    2. Profitability Metrics:
    Return on Equity: {metrics.get('roeRfy', 'N/A')}%
    - Interpretation: {'Excellent' if metrics.get('roeRfy', 0) > 20 else 'Good' if 15 <= metrics.get('roeRfy', 0) <= 20 else 'Average' if 10 <= metrics.get('roeRfy', 0) < 15 else 'Poor'}

    Net Profit Margin: {metrics.get('netProfitMarginTTM', 'N/A')}%
    - Interpretation: {'Excellent' if metrics.get('netProfitMarginTTM', 0) > 20 else 'Good' if 10 <= metrics.get('netProfitMarginTTM', 0) <= 20 else 'Average' if 5 <= metrics.get('netProfitMarginTTM', 0) < 10 else 'Poor'}

    3. Liquidity and Solvency:
    Current Ratio: {metrics.get('currentRatioQuarterly', 'N/A')}
    - Interpretation: {'Strong' if metrics.get('currentRatioQuarterly', 0) > 2 else 'Healthy' if 1.5 <= metrics.get('currentRatioQuarterly', 0) <= 2 else 'Adequate' if 1 <= metrics.get('currentRatioQuarterly', 0) < 1.5 else 'Poor'}

    Debt-to-Equity Ratio: {metrics.get('totalDebtToEquityQuarterly', 'N/A')}
    - Interpretation: {'Low leverage' if metrics.get('totalDebtToEquityQuarterly', 0) < 0.5 else 'Moderate leverage' if 0.5 <= metrics.get('totalDebtToEquityQuarterly', 0) <= 1 else 'High leverage'}

    4. Dividend Analysis:
    Dividend Yield: {metrics.get('dividendYieldIndicatedAnnual', 'N/A')}%
    - Interpretation: {'High yield' if metrics.get('dividendYieldIndicatedAnnual', 0) > 4 else 'Moderate yield' if 2 <= metrics.get('dividendYieldIndicatedAnnual', 0) <= 4 else 'Low yield'}

    5. Market Performance:
    52-Week Range: ${metrics.get('52WeekLow', 'N/A')} - ${metrics.get('52WeekHigh', 'N/A')}
    Current Price Position: {((price_data['current_price'] - metrics.get('52WeekLow', price_data['current_price'])) / (metrics.get('52WeekHigh', price_data['current_price']) - metrics.get('52WeekLow', price_data['current_price'])) * 100):.2f}% of 52-Week Range

    Beta: {metrics.get('beta', 'N/A')}
    - Interpretation: {'More volatile than market' if metrics.get('beta', 1) > 1 else 'Less volatile than market' if metrics.get('beta', 1) < 1 else 'Same volatility as market'}

    Overall Analysis:
    {profile['name']} shows {'strong' if metrics.get('roeRfy', 0) > 15 and metrics.get('currentRatioQuarterly', 0) > 1.5 else 'moderate' if metrics.get('roeRfy', 0) > 10 and metrics.get('currentRatioQuarterly', 0) > 1 else 'weak'} financial health with {'high' if metrics.get('peBasicExclExtraTTM', 0) > 25 else 'moderate' if 15 <= metrics.get('peBasicExclExtraTTM', 0) <= 25 else 'low'} valuation metrics. The company's profitability is {'excellent' if metrics.get('netProfitMarginTTM', 0) > 20 else 'good' if metrics.get('netProfitMarginTTM', 0) > 10 else 'average' if metrics.get('netProfitMarginTTM', 0) > 5 else 'poor'}, and it has {'low' if metrics.get('totalDebtToEquityQuarterly', 0) < 0.5 else 'moderate' if metrics.get('totalDebtToEquityQuarterly', 0) < 1 else 'high'} financial leverage. Investors should consider these factors along with their investment goals and risk tolerance.


    Recent News and Sentiment Analysis:
    """

    for item in data["sentiments"]:
        report += f"""
    Title: {item['title']}
    URL: {item['url']}
    Sentiment Analysis: {item['sentiment']} (Confidence: {item['confidence']:.2f})

    """
    # Content Preview: {item['content'][:500]}...
    return report


class Tools:
    class UserValves(BaseModel):
        FINNHUB_API_KEY: str = ""
        pass

    def __init__(self):
        pass

    async def compile_stock_report(
        self,
        ticker: str,
        __user__: dict = {},
        __event_emitter__: Callable[[Any], Awaitable[None]] = None,
    ) -> str:
        """
        Perform a comprehensive stock analysis and compile a detailed report for a given ticker using Finnhub's API.

        This function gathers various data points including:
        - Basic company information (industry, market cap, etc.)
        - Current trading information (price, daily change, etc.)
        - Key financial metrics (P/E ratio, EPS, ROE, etc.)
        - List of peer companies
        - Recent news articles with sentiment analysis using FinBERT

        The gathered data is then compiled into a structured, easy-to-read report.

        :param ticker: The stock ticker symbol (e.g., "AAPL" for Apple Inc.).
        :return: A comprehensive analysis report of the stock as a formatted string.
        """
        await __event_emitter__(
            {
                "type": "status",
                "data": {"description": "Initializing client", "done": False},
            }
        )
        self.client = finnhub.Client(api_key=__user__["valves"].FINNHUB_API_KEY)
        await __event_emitter__(
            {
                "type": "status",
                "data": {"description": "Retrieving stock data", "done": False},
            }
        )
        data = await _async_gather_stock_data(self.client, ticker)
        await __event_emitter__(
            {
                "type": "status",
                "data": {"description": "Compiling stock report", "done": False},
            }
        )
        report = _compile_report(data)
        # Get lastest price from data
        last_price = data["current_price"]["current_price"]
        await __event_emitter__(
            {
                "type": "status",
                "data": {
                    "description": "Finished creating report - latest price: "
                    + str(last_price),
                    "done": True,
                },
            }
        )
        return report

**Comment by SignedAdam** [comment]
How did you get to use the tools at all? Could you give an example prompt? I am having trouble tools at all with any model

**Reply by SignedAdam** [reply]
update: installing the latest version (0.3.6) makes the tools work some of the times with LLama 3 8b Instruct (and better models)

**Reply by aleixdorca** [reply]
Hey @SignedAdam,

I guess by your update that you solved how to use tools.

This is how I do it:

- Admin User -> Workspace -> Tools -> Add tool
- Add a name and description
- Add the python code to execute
- Then, go to the Models tab, and enable the tools for a particular model
- Now tools are available at the chat, when pressing the '+' button

To debug the tools check the open-webui logs. In my case, I use docker and the command is: $ docker logs <container_id> -f

You will see that when tools are called the relevant json strings appear on the logs. This, to me, is the crucial part. The way I see it, from the user prompt a call to the current LLM is made to choose the tool and create the json that will have the parameters for the python function you have created in the Tools section. The content this call return may not always be accurate. :sad: Each model replies as it feels. Maybe temperature should be set to a low value, here. I don't know.

Hope it helps.

**Reply by SignedAdam** [reply]
Thank you for your elaborate reply.

I did manage to fix it, but this is very helpful nonetheless as I am also trying to create a few tools.

In my case, 'llama 3 8b instruct' manages to call the right tool with the right parameters most of the time. 

This is of interest to me and if you would like to connect we can analyze and solve it together.

A few notes that may or may not be important:
- I use the models over OpenRouter. I suspect they are fp16 (non quantized as opposed to models commonly used locally with ollama)
- I had to update openwebui to the latest version, or the tools would not work at all
- I am using 'llama 3 8b instruct'
- my tools are very simple, with only 1 input parameter (I will post an example when I get on my computer)

**Reply by SignedAdam** [reply]
got caught up in some stuff. Here is my example Tool.

```python
"""
title: YouTube Video Downloader
author: Assistant
version: 0.1.0
"""

import os
import requests
from typing import Dict, Any
from pydantic import BaseModel, Field


class Tools:
    class Valves(BaseModel):
        RAPIDAPI_HOST: str = Field(
            default="yt-api.p.rapidapi.com",
            description="The RapidAPI host for the YouTube Downloader API.",
        )
        RAPIDAPI_KEY: str = Field(
            default="set_me",
            description="Your RapidAPI API key.",
        )

    def __init__(self):
        self.citation = True
        self.valves = self.Valves()
        self.download_dir = "/app/backend/data/downloads"

    def download_video(self, video_url: str) -> str:
        """
        Download a YouTube video using the RapidAPI YouTube Downloader API and save it to a local directory.

        Args:
        video_url (str): The URL of the YouTube video to download.

        Returns:
        str: A message indicating the download status and the path of the downloaded video.
        """
        try:
            # Create the download directory if it doesn't exist
            os.makedirs(self.download_dir, exist_ok=True)

            # Extract video ID from various YouTube URL formats
            if "youtu.be/" in video_url:
                video_id = video_url.split("youtu.be/")[1].split("?")[0]
            elif "youtube.com/watch?v=" in video_url:
                video_id = video_url.split("v=")[1].split("&")[0]
            elif "youtube.com/embed/" in video_url:
                video_id = video_url.split("embed/")[1].split("?")[0]
            elif "youtube.com/v/" in video_url:
                video_id = video_url.split("v/")[1].split("?")[0]
            elif "youtube.com/shorts/" in video_url:
                video_id = video_url.split("shorts/")[1].split("?")[0]
            else:
                raise ValueError("Invalid YouTube URL format")
            url = f"https://{self.valves.RAPIDAPI_HOST}/dl?id={video_id}"

            headers = {
                "X-RapidAPI-Host": self.valves.RAPIDAPI_HOST,
                "X-RapidAPI-Key": self.valves.RAPIDAPI_KEY,
            }

            response = requests.get(url, headers=headers)
            response.raise_for_status()

            data = response.json()
            # Extract the highest quality video format
            formats = data.get("formats", [])
            video_format = max(
                formats, key=lambda x: x.get("height", 0) if x.get("height") else 0
            )
            download_url = video_format.get("url")

            if download_url:
                video_response = requests.get(download_url, stream=True)
                video_response.raise_for_status()

                video_title = data.get("title", "video")
                safe_title = "".join(
                    [c for c in video_title if c.isalnum() or c in (" ", "-", "_")]
                ).rstrip()
                video_path = os.path.join(
                    self.download_dir, f"{safe_title}_{video_id}.mp4"
                )

                with open(video_path, "wb") as video_file:
                    for chunk in video_response.iter_content(chunk_size=8192):
                        if chunk:
                            video_file.write(chunk)

                return f"Video '{safe_title}' successfully downloaded and saved to: {video_path}"
            else:
                return "Failed to retrieve download URL."

        except requests.exceptions.RequestException as e:
            return f"An error occurred while downloading the video: {str(e)}"
        except Exception as e:
            return f"An unexpected error occurred lol: {str(e)}"
```
This seems to work for me 10 times out of 10, with llama 3 8b (instruct)

**Reply by aleixdorca** [reply]
Adam,

I have correctly adapted your code to install it as a tool in my instance.

The problem is the same though. It has nothing to do with the tool in itself. Gemma 2 keeps adding the ```json [tool parameters] ``` delimiters when preparing the call.

Most of the time, the call fails as shown in the logs:

```
['calculator']
calculator
History:
USER: """calculate the following: 2+3+4*2-10"""
SYSTEM: """You are a helpful assistant"""
Query: calculate the following: 2+3+4*2-10
generate_ollama_chat_completion
content: ```json
{
 "name": "calculator",
 "parameters": {
  "equation": "2+3+4*2-10"
 }
}```
Error: Expecting value: line 1 column 1 (char 0)
False
tool_contexts: []
```

No worries, though, I know how to prevent it using the regex provided above, so it's no big deal.

Thanks for the interest.
